siwe.go

```diff
@@ -0,0 +1,462 @@
+package siwe
+
+import (
+	"crypto/ecdsa"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/dchest/uniuri"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/relvacode/iso8601"
+)
+
+type MalformedMessage struct{ string }
+type ParsingFailed struct{ string }
+type ExpiredMessage struct{ string }
+type InvalidMessage struct{ string }
+type InvalidSignature struct{ string }
+
+func (m *ParsingFailed) Error() string {
+	return "Expired Message"
+}
+
+func (m *ExpiredMessage) Error() string {
+	return "Expired Message"
+}
+
+func (m *InvalidMessage) Error() string {
+	return "Invalid Message"
+}
+
+func (m *InvalidSignature) Error() string {
+	return fmt.Sprintf("Invalid Signature: %s", m.string)
+}
+
+type Message struct {
+	domain  string
+	address common.Address
+	uri     string
+	version string
+
+	statement *string
+	nonce     *string
```

Comment:
I think it might be preferable to not have it optional, if a user wants to build a message they should create a nonce and be forced to store it somewhere for verification later.

Reply:
That was my mistake on typing, it's a pointer, but it's not actually optional, in the case the user doesn't provide one, one will be automatically generated and added to the message. I see now that this might prevent an user from verifying a message that doesn't have a `nonce`, since it's impossible to construct one without with this library.
We can either:
- Keep adding a nonce if not present in the initialization options; or
- Throw an error if not present.
Besides that, I've fixed the type in the review branch.

Reply:
I think probably throw an error. I suppose ideally we wouldn't have a builder method and just let the user populate the struct, but it could be an issue due to loose type.